File: sdk/desktop/macos/Tests/KsealDesktopTests/HardwareKeyStoreTests.swift
import XCTest
@testable import KsealDesktop

final class HardwareKeyStoreTests: XCTestCase {

    private func tempDir() -> URL {
        let dir = FileManager.default.temporaryDirectory
            .appendingPathComponent("kseal-hwkey-\(UUID().uuidString)", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    private func proofFile(_ dir: URL) -> URL {
        dir.appendingPathComponent("kseal", isDirectory: true).appendingPathComponent("proof.key")
    }

    func testFakeStoreRoundTrips() throws {
        let store = FakeHardwareKeyStore()
        let secret = Data([1, 2, 3, 4, 5, 250, 251, 252])
        let sealed = try store.seal(secret)
        XCTAssertNotEqual(sealed, secret)
        XCTAssertEqual(try store.unseal(sealed), secret)
    }

    func testUnsealRejectsForeignBlob() {
        let store = FakeHardwareKeyStore()
        XCTAssertThrowsError(try store.unseal(Data([0, 1, 2, 3, 4, 5])))
    }

    func testSoftwareStoreIsPassthroughAndMatchesLegacyLayout() throws {
        let dir = tempDir()
        let provider = HardwareBoundProofKeyProvider(directory: dir, store: SoftwareKeyStore())
        XCTAssertFalse(provider.isHardwareBacked)

        let key = provider.proofKey()
        XCTAssertEqual(key.count, HardwareBoundProofKeyProvider.keyLength)
        // Software fallback persists the raw key verbatim (legacy on-disk layout).
        let onDisk = try Data(contentsOf: proofFile(dir))
        XCTAssertEqual(onDisk, key)

        // Stable across instances pointing at the same storage.
        let again = HardwareBoundProofKeyProvider(directory: dir, store: SoftwareKeyStore()).proofKey()
        XCTAssertEqual(again, key)
    }

    func testHardwareStoreSealsAtRestButKeyIsStable() throws {
        let dir = tempDir()
        let store = FakeHardwareKeyStore()
        let provider = HardwareBoundProofKeyProvider(directory: dir, store: store)
        XCTAssertTrue(provider.isHardwareBacked)

        let key = provider.proofKey()
        XCTAssertEqual(key.count, HardwareBoundProofKeyProvider.keyLength)

        // What is persisted is the sealed blob, never the raw key.
        let onDisk = try Data(contentsOf: proofFile(dir))
        XCTAssertNotEqual(onDisk, key)
        XCTAssertEqual(try store.unseal(onDisk), key)

        // A second provider over the same store + storage yields the same key.
        let again = HardwareBoundProofKeyProvider(directory: dir, store: FakeHardwareKeyStore()).proofKey()
        XCTAssertEqual(again, key)
    }

    func testLegacyRawKeyIsMigratedToSealedInPlace() throws {
        let dir = tempDir()
        // Simulate a pre-hardware install: a raw 32-byte key on disk.
        let legacy = Data((0..<32).map { UInt8($0) })
        let ksealDir = dir.appendingPathComponent("kseal", isDirectory: true)
        try FileManager.default.createDirectory(at: ksealDir, withIntermediateDirectories: true)
        try legacy.write(to: proofFile(dir))

        let store = FakeHardwareKeyStore()
        let provider = HardwareBoundProofKeyProvider(directory: dir, store: store)
        let key = provider.proofKey()

        // Continuity: the migrated key equals the legacy key.
        XCTAssertEqual(key, legacy)
        // And the on-disk blob is now sealed (no longer the raw legacy bytes).
        let onDisk = try Data(contentsOf: proofFile(dir))
        XCTAssertNotEqual(onDisk, legacy)
        XCTAssertEqual(try store.unseal(onDisk), legacy)
    }

    func testSealFailureDegradesToSoftwareKeyWithoutCrashing() throws {
        let dir = tempDir()
        let provider = HardwareBoundProofKeyProvider(directory: dir, store: FailingHardwareKeyStore())
        let key = provider.proofKey()
        XCTAssertEqual(key.count, HardwareBoundProofKeyProvider.keyLength)
        // Persisted (raw, since sealing failed) and stable on reload.
        let onDisk = try Data(contentsOf: proofFile(dir))
        XCTAssertEqual(onDisk, key)
    }

    func testDefaultHardwareKeyStoreIsConstructible() {
        // On Linux/CI this is the software fallback; on macOS a Secure-Enclave
        // store when available. Either way it must produce a usable store.
        let store = makeDefaultHardwareKeyStore(label: "tenant-app")
        let secret = Data(repeating: 7, count: 32)
        if let sealed = try? store.seal(secret) {
            XCTAssertEqual(try? store.unseal(sealed), secret)
        }
    }
}

File: sdk/desktop/windows/tests/Kseal.Desktop.Tests/HardwareKeyStoreTests.cs
using Xunit;

namespace Kseal.Desktop.Tests;

public class HardwareKeyStoreTests : IDisposable
{
    private readonly string _dir = Path.Combine(Path.GetTempPath(), $"kseal-hw-{Guid.NewGuid():N}");

    public HardwareKeyStoreTests() => Directory.CreateDirectory(_dir);

    public void Dispose()
    {
        try { Directory.Delete(_dir, recursive: true); } catch (IOException) { }
        GC.SuppressFinalize(this);
    }

    private string KeyPath => Path.Combine(_dir, "kseal", "proof.key");

    [Fact]
    public void SoftwareStorePassesThroughUnchanged()
    {
        var store = new SoftwareKeyStore();
        byte[] data = [1, 2, 3, 4];
        Assert.False(store.IsHardwareBacked);
        Assert.Equal(data, store.Seal(data));
        Assert.Equal(data, store.Unseal(data));
    }

    [Fact]
    public void FakeStoreRoundTripsAndRejectsForeignBlob()
    {
        var store = new FakeHardwareKeyStore();
        byte[] data = [9, 8, 7, 6, 5];
        byte[] sealed_ = store.Seal(data);
        Assert.NotEqual(data, sealed_);                 // actually sealed
        Assert.Equal(data, store.Unseal(sealed_));      // round-trips
        Assert.Throws<HardwareKeyStoreException>(() => store.Unseal([0, 1, 2])); // foreign blob
    }

    [Fact]
    public void HardwareBoundProviderSealsKeyAtRestAndIsStable()
    {
        var store = new FakeHardwareKeyStore();
        var provider = new HardwareBoundProofKeyProvider(_dir, store);

        byte[] key1 = provider.ProofKey();
        Assert.Equal(32, key1.Length);
        Assert.True(provider.IsHardwareBacked);

        // The on-disk blob is sealed, not the raw key.
        byte[] onDisk = File.ReadAllBytes(KeyPath);
        Assert.NotEqual(key1, onDisk);
        Assert.Equal(key1, store.Unseal(onDisk));

        // Stable across provider instances (re-reads + unseals the same key).
        byte[] key2 = new HardwareBoundProofKeyProvider(_dir, store).ProofKey();
        Assert.Equal(key1, key2);
    }

    [Fact]
    public void LegacyRawKeyIsAdoptedAndResealedInPlace()
    {
        // Simulate a pre-hardware install: a raw 32-byte key already on disk.
        Directory.CreateDirectory(Path.GetDirectoryName(KeyPath)!);
        byte[] legacy = new byte[32];
        for (int i = 0; i < 32; i++) legacy[i] = (byte)(i + 1);
        File.WriteAllBytes(KeyPath, legacy);

        var store = new FakeHardwareKeyStore();
        var provider = new HardwareBoundProofKeyProvider(_dir, store);

        byte[] key = provider.ProofKey();
        Assert.Equal(legacy, key); // continuity: same key

        // It was re-sealed in place, so the on-disk bytes are now the sealed blob.
        byte[] onDisk = File.ReadAllBytes(KeyPath);
        Assert.NotEqual(legacy, onDisk);
        Assert.Equal(legacy, store.Unseal(onDisk));
    }

    [Fact]
    public void SealFailureDegradesToRawKeyWithoutBricking()
    {
        var provider = new HardwareBoundProofKeyProvider(_dir, new FailingHardwareKeyStore());
        byte[] key1 = provider.ProofKey();
        Assert.Equal(32, key1.Length);
        // Persisted as raw (software-equivalent) and stable.
        Assert.Equal(key1, File.ReadAllBytes(KeyPath));
        byte[] key2 = new HardwareBoundProofKeyProvider(_dir, new FailingHardwareKeyStore()).ProofKey();
        Assert.Equal(key1, key2);
    }

    [Fact]
    public void DefaultFactoryProducesAUsableStore()
    {
        // On non-Windows CI this is the software fallback; on Windows w/o TPM too.
        IHardwareKeyStore store = HardwareKeyStoreFactory.Create("tenant-app");
        byte[] data = [1, 2, 3];
        Assert.Equal(data, store.Unseal(store.Seal(data)));
    }
}
